support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,342 @@
+
+		r.forEachPodEndpointAddress(&endpoints, func(address corev1.EndpointAddress) {
+			if pod.Name == address.TargetRef.Name && pod.Namespace == address.TargetRef.Namespace {
+				// 2.4.1: <hostname>.<service>.<ns>.svc.<zone>
+				if address.Hostname != "" {
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.Hostname, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				} else {
+					// The spec leaves this case up to the implementation, so here we copy CoreDns...
+					// CoreDNS has an endpoint_pod_names flag to switch between the following two options. We don't have that flag, so
+					// we'll just add both pod name and IP based name (the CoreDns default) for now.
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", r.mungeIP(address.IP), endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.TargetRef.Name, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				}
+			}
+		})
+	}
+
+	namesSlice := make([]string, len(names))
+	i := 0
+	for name := range names {
+		namesSlice[i] = name
+		i++
+	}
+
+	return namesSlice, nil
+}
+
+func (r *PodReconciler) fillEntryForPod(ctx context.Context, entry *common.RegistrationEntry, pod *corev1.Pod) (*common.RegistrationEntry, error) {
+	if !r.AddPodDNSNames {
+		return entry, nil
+	}
+
+	if pod.Status.PodIP == "" {
+		// Pod doesn't have an IP yet, so we can't generate names properly yet. We'll be called to reconcile it again
+		// once it has an IP, at which point we'll fill the entry.
+		return entry, nil
+	}
+
+	endpointNames, err := r.getNamesForEndpoints(ctx, pod)
+	if err != nil {
+		return nil, err
+	}
+
+	// We want to add DNS names for our pod. The order of these is important, as the first one will be used as the DN
+	// for X509 SVIDs. This means we would like to use a completely unambiguous, unique, name for our pod first.
+	// The docs at https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods claim that we should have:
+	// `pod-ip-address.my-namespace.pod.cluster-domain.example` for all pods.
+	// CoreDns does indeed have a pods option allowing resolution names in the form 1-2-3-4.ns.pod.cluster.local.
+	// It's enabled only if the CoreDns pods option is set to a value other than `disabled.` We will include this as the
+	// first dns name (regardless of configuration) as it makes for a good DN.
+
+	// The docs also claim we should have a `pod-ip-address.deployment-name.my-namespace.svc.cluster-domain.example`
+	// name for daemonsets/deployments exposed by a service. This doesn't seem to be implemented by anything, and would
+	// clash with naming used for endpoints! We follow CoreDns here, and don't attempt to add that format of name.
+
+	entry.DnsNames = make([]string, len(endpointNames)+1)
+	entry.DnsNames[0] = fmt.Sprintf("%s.%s.pod.%s", r.mungeIP(pod.Status.PodIP), pod.Namespace, r.ClusterDNSZone)
+	for i := range endpointNames {
+		entry.DnsNames[i+1] = endpointNames[i]
+	}
```

Comment:
I can't imagine any sort of allocation optimization to be important here, I think the following is much simpler:
```suggestion
	entry.DnsNames = append([]string{fmt.Sprintf("%s.%s.pod.%s", r.mungeIP(pod.Status.PodIP), pod.Namespace, r.ClusterDNSZone)}, endpointNames...)
```